src/print_graph.cc

```diff
@@ -37,120 +35,136 @@ namespace detail {
 	}
 
 	std::string print_task_graph(const task_ring_buffer& tdag) {
-		std::ostringstream ss;
-		ss << "digraph G { label=\"Task Graph\" ";
+		std::string dot = "digraph G {label=\"Task Graph\" ";
+		auto dot_append = std::back_inserter(dot);
 
 		for(auto tsk : tdag) {
-			std::unordered_map<std::string, std::string> props;
-			props["label"] = "\"" + get_task_label(tsk) + "\"";
-
-			ss << tsk->get_id();
-			ss << "[";
-			for(const auto& it : props) {
-				ss << " " << it.first << "=" << it.second;
-			}
-			ss << "];";
-
+			fmt::format_to(dot_append, "{}[label=\"{}\"];", tsk->get_id(), get_task_label(tsk));
 			for(auto d : tsk->get_dependencies()) {
-				ss << fmt::format("{} -> {} [{}];", d.node->get_id(), tsk->get_id(), dependency_style(d));
+				fmt::format_to(dot_append, "{}->{}[{}];", d.node->get_id(), tsk->get_id(), dependency_style(d));
 			}
 		}
 
-		ss << "}";
-		return ss.str();
+		dot += "}";
+		return dot;
 	}
 
-	std::string get_command_label(const abstract_command* cmd) {
-		std::string label = fmt::format("[{}] Node {}:\\n", cmd->get_cid(), cmd->get_nid());
-		if(const auto ecmd = dynamic_cast<const epoch_command*>(cmd)) {
+	std::string get_command_label(const abstract_command& cmd, const task_manager& tm, const reduction_manager& rm) {
+		const command_id cid = cmd.get_cid();
+		const node_id nid = cmd.get_nid();
+
+		std::string label = fmt::format("[{}] Node {}:\\n", cid, nid);
```

Comment:
Maybe we should use this chance to make these prefixes a bit shorter (something like `{} N{}\\n`).

Reply:
Done.